tac-operation-lifecycle/tac-operation-lifecycle-logic/src/database.rs

```diff
@@ -0,0 +1,1167 @@
+                        chunk.len(),
+                        if let Some(wm) = updated_watermark {
+                            wm.to_string()
+                        } else {
+                            "[NOT_UPDATED]".to_string()
+                        },
+                    );
+                }
+                Err(e) => {
+                    tx.rollback().await?;
+                    tracing::error!("Failed to save batch: {}", e);
+                    return Err(e.into());
+                }
+            }
+        }
+
+        Ok(intervals.len())
+    }
+
+    // stage_types is a vector Id->StageName
+    pub async fn register_stage_types(
+        &self,
+        stage_types: &HashMap<i32, String>,
+    ) -> anyhow::Result<()> {
+        for (stage_id, stage_name) in stage_types.into_iter() {
+            let stage_type_model = stage_type::ActiveModel {
+                id: Set(*stage_id),
+                name: Set(stage_name.to_string()),
+            };
+
+            stage_type::Entity::insert(stage_type_model)
+                .on_conflict(
+                    OnConflict::column(stage_type::Column::Id)
+                        .update_columns([stage_type::Column::Name])
+                        .to_owned(),
+                )
+                .exec(self.db.as_ref())
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    pub async fn insert_pending_operations(
+        &self,
+        operations: &ApiOperations,
+    ) -> anyhow::Result<()> {
+        let thread_id = thread::current().id();
+
+        // Start a transaction
+        let txn = self.db.begin().await?;
+
+        // Save all operations
+        for op in operations {
+            let operation_model = operation::ActiveModel {
+                id: Set(op.id.clone()),
+                operation_type: Set(None),
+                timestamp: Set(op.timestamp as i64),
+                status: Set(EntityStatus::Pending.to_id() as i32),
+                next_retry: Set(None),
+                retry_count: Set(0), // Initialize retry count
+            };
+
+            tracing::debug!(
+                "[Thread {:?}] Attempting to insert operation: {:?}",
+                thread_id,
+                operation_model
+            );
+
+            // Use on_conflict().do_nothing() with proper error handling
+            match operation::Entity::insert(operation_model)
+                .on_conflict(
+                    sea_orm::sea_query::OnConflict::column(operation::Column::Id)
+                        .do_nothing()
+                        .to_owned(),
+                )
+                .exec(&txn)
+                .await
+            {
+                Ok(_) => tracing::debug!(
+                    "[Thread {:?}] Successfully inserted or skipped operation",
+                    thread_id
+                ),
+                Err(e) => {
+                    tracing::debug!(
```

Comment:
maybe warn?

Reply:
or is it not important?

Reply:
Good point, thank you. I divided it into the two cases (inserted or DO NOTHING on conflic)  in https://github.com/blockscout/blockscout-rs/pull/1325/commits/c024b40c2dc4f2da193b7e375f6ffc1a4c38633c